lib/ansible/modules/cloud/azure/azure_vm_metadata_facts.py

```diff
@@ -0,0 +1,369 @@
+        encoded_url = quote(url, safe='%/:=&?~#+!$,;\'@()*[]')
+        response, info = fetch_url(self.module, encoded_url, method="GET", force=True, headers={'Metadata': 'true'})
+        if info['status'] not in (200, 404):
+            time.sleep(3)
+            #
+            # Request failed.
+            # Retry the request then fail...
+            #
+            self.module.warn('Retrying query to metadata service. First attempt failed: {0}'.format(info['msg']))
+            response, info = fetch_url(self.module, encoded_url, method="GET", force=True, headers={'Metadata': 'true'})
+            if info['status'] not in (200, 404):
+                #
+                # Second request failed.
+                # Failure.
+                #
+                self.module.warn('Second query to metadata service failed: {0}'.format(info['msg']))
+                self.module.fail_json(msg='Failed to retrieve metadata from Azure: {0}'.format(info['msg']), response=info)
+        if response:
+            data = response.read()
+        else:
+            data = None
+        return to_text(data)
+
+
+    def _keyfix(self, fields):
+
+        for key, val in fields.items():
+            #
+            # Change ':'' and '-' to '_' to ensure valid template variable names
+            #
+            newkey = re.sub(':|-', '_', key).lower()
+
+            fields[newkey] = fields.pop(key)
+            #
+            # if list, roll through the list and parse what is inside
+            #
+            if isinstance(val, list):
+                templist = []
+                for l in val:
+                    templist.append(self._keyfix(l))
+                fields[newkey] = templist
+            #
+            # if dictionary, recurse through
+            #
+            elif isinstance(val, dict):
+                fields[newkey] = self._keyfix(val)
+            else:
+                fields[newkey] = val
+
+        return fields
+
+
+    #
+    # Designed to create template-friendly variable names.
+    # While keyfix just returns a dictionary, this function
+    # seperates every key+value in to a variable.
+    #
+    def _json_to_facts(self, fields, prefix_key=""):
+
+        if (not prefix_key):
+            prefix_key = (self._prefix % "_")
+
+        for key, val in fields.items():
+
+            #
+            # Change ':'' and '-' to '_' to ensure valid template variable names
+            #
+            newkey = prefix_key + re.sub(':|-', '_', key).lower()
+            #
+            # if list, roll through the list and parse what is inside
+            #
+            if isinstance(val, list):
+                for l in val:
+                    self._json_to_facts(l, str(newkey) + "_")
+            #
+            # if dictionary, recurse through
+            #
+            elif isinstance(val, dict):
+                self._json_to_facts(val, str(newkey) + "_")
+            else:
+                self._data[newkey] = val
```

Comment:
construct a local dict in this method and add result to to this dict, finally return it. DON'T modified a complicated global variable in different methods, this makes the code hard to read and maintain.